SCons: Fix handling of platform-specific tools, notably `mingw`

akien-mga opened the pull request:
- Fixes #101186

Redo of #99762 on top of #101715, aiming to fix various issues we have with the fact that we need to know the selected platform to define the right SCons "tools" to enable when creating the environment, but we need the environment to register command line options that help set said platform.

Using two environments with `env.Clone()` seems to help solve the problems in #101715, and fixing the detection of when `mingw` should be set as tool via #99762 seems to make it work out of the box for Linux+mingw at least.

akien-mga commented on SConstruct:
```diff
@@ -430,6 +415,18 @@ env.modules_detected = modules_detected
+    custom_tools = detect.get_tools(env)
+except AttributeError:
+    pass
+env = env.Clone(tools=custom_tools)
```
This fixes the problem of requiring an `env` to parse options, but requiring options to be parsed to register the env with the right tools - at least in my early testing only for Windows mingw builds on Linux.

I haven't read much about what `Environment.Clone()` actually does so we need to do more research to confirm that it's the right tool for the job.

Repiteo replied:
This shouldn't be necessary if using the above empty tool argument.
```suggestion
for tool in custom_tools:
	env.Tool(tool)
```

akien-mga replied:
Interesting! Actually I think the `Clone()` wasn't required either way, it also seems to work fine with the original `Environment(tools=["default"])`.

The issue in https://github.com/godotengine/godot/pull/101715/files#r1920844908 was due to `mingw` not being properly set as custom tool when just doing `scons p=windows` on Linux, because our logic for that was incomplete. Why it worked before our PRs though, I'm not sure.

akien-mga replied:
> The issue in https://github.com/godotengine/godot/pull/101715/files#r1920844908 was due to `mingw` not being properly set as custom tool when just doing `scons p=windows` on Linux, because our logic for that was incomplete. Why it worked before our PRs though, I'm not sure.

Actually I think the reason it failed in #101715 is because of the removal of `default`.

So I'm still unsure whether it's a good idea to do `tools=[]` and then register some custom tools, or `tools=["default"]` to have sane defaults and then add custom tools (like we currently do in `master`).

akien-mga replied:
> Actually I think the reason it failed in #101715 is because of the removal of `default` here:

Or not, there must be something more linked with the `Environment()` constructor.

akien-mga commented on SConstruct:
```diff
@@ -340,21 +340,6 @@ if env["platform"] not in platform_list:
-elif env["platform"] == "windows" and methods.get_cmdline_bool("use_mingw", False):
-    custom_tools += ["mingw"]
-for tool in custom_tools:
-    env.Tool(tool)
```
I moved this further down because there's more options overrides below, notably platform flags, and we likely want to respect them before calling `detect.get_tools(env)` so that this can make full use of the configuration so far.

Repiteo commented on SConstruct:
```diff
@@ -107,22 +107,10 @@ for x in sorted(glob.glob("platform/*")):
 # want to have to pull in manually.
 # Then we prepend PATH to make it take precedence, while preserving SCons' own entries.
-env = Environment(tools=custom_tools)
+env = Environment()
```
While testing, I realized that SCons defaults to "default" as the tool passed if no tool argument is given. So this should be an empty list, which *should* remove the need for the `env.Clone` workaround
```suggestion
env = Environment(tools=[])
```

akien-mga commented on SConstruct:
```diff
@@ -434,6 +415,22 @@ env.modules_detected = modules_detected
+# FIXME: Tool assignment happening at this stage is a direct consequence of getting the platform logic AFTER the SCons
+# environment was already been constructed. Fixing this would require a broader refactor where all options are setup
+# ahead of time with native validator/converter functions.
+tmppath = "./platform/" + env["platform"]
+sys.path.insert(0, tmppath)
+import detect
```
This had to move up, otherwise `detect` here was actually the one of the last platform we checked when parsing all of them... i.e. windows, so Linux and macOS were getting `mingw` registered and ending with `.exe` binary extensions :p

Repiteo reviewed (approved):
Gonna do more thorough testing later, but everything appears to work-as-expected on Windows! It'll be pretty handy to have this specialized per-platform, and future PRs being able to trim that fat even further is an exciting prospect

EDIT: WSL validated & no longer causes overrides! Removing draft status

Calinou reviewed (approved):
Tested locally on Fedora 41 with MinGW, it works as expected.

bruvzg reviewed (approved):
Tested on macOS (GCC and clang toolchains) and Windows (standalone MinGW and MSYS2 GCC), seems to be working fine.

akien-mga committed:
SCons: Fix handling of platform-specific tools, notably `mingw`

Add optional `detect.py` `get_tools` method to let platforms register SCons
tools they need.

This helps move this logic out of SConstruct, keeping platforms more self
contained, and helping thirdparty platforms define their own custom tools.

This logic was also unreliable (the `use_mingw` one would only work if
passed manually on the command line, not in e.g. `get_flags`).

Co-authored-by: Thaddeus Crews <repiteo@outlook.com>

akien-mga commented:
Rebased and squashed.

Repiteo commented:
Thanks!
